### PR TITLE
Make `argnums` in `jacrev` and `jacfwd` truely optional.

### LINKED ISSUE #118861: Wrong typing for `argnums` in `torch.func.jacrev`, `torch.func.jacfwd` and possibly `torch.func.vmap`.
### 📚 The doc issue

It seems that in the code and in the docs, the type of `argnums` in `torch.func.jacrev` is supposed to be `Union[int, Tuple[int, ...]]` and it is actually `Union[int, Tuple[int]]`. I think this also holds for `jacfwd` and possibly for the parameter `output_dims` of `vmap`.

For `jacrev` : 
- The docs : https://pytorch.org/docs/stable/generated/torch.func.jacrev.html
- The code : https://github.com/pytorch/pytorch/blob/eaa45f47f825777599c097af577d854bb47711ae/torch/_functorch/eager_transforms.py#L367

### Suggest a potential alternative/fix

I believe all those `Tuple[int]` should be changed to `Tuple[int, ...]`.

### PR DESCRIPTION
**Pull request by PierreQuinton**
Fixes #118861

For both `jacrev` and `jacfwd`:
- Fixes the typing of `argnums` to `Optional[argnums_t]`.
- Improves the doc to specify behavior in case of `None`, in this case we take the jacobian w.r.t. all inputs.
- Tests the `None` parameter.

I was able to lint the PR, however I could not run tests despite my best efforts to make it work. The only test that can fail should be the one I wrote.

As a side comment, it feels like the default value for `argnums` in both `jacrev` and `jacfwd` would be more naturally `None` instead of `0`. The problem is that it would make the change not backward compatible and in particular the typical use of default argument is when we have only one argument to differentiate with respect to it but in that case we would return a tuple containing only one tensor instead of the said tensor. Therefore it feels hard to make that change.

### CODE REVIEW

File: torch/_functorch/eager_transforms.py [hunk 1]
```diff
@@ -364,7 +364,7 @@ def error_if_complex(func_name, args, is_input):
             raise RuntimeError(err_msg)
 
 @exposed_in("torch.func")
-def jacrev(func: Callable, argnums: Union[int, Tuple[int]] = 0, *, has_aux=False,
+def jacrev(func: Callable, argnums: Optional[argnums_t] = 0, *, has_aux=False,
```
Comment (zou3519) [inline torch/_functorch/eager_transforms.py:367]: I wrote this on the issue, but argnums cannot actually be None, it must be int or Tuple[int, ...]. If we accept `None` here then that's a bug (and we should add a check against it)

Comment (PierreQuinton) [reply torch/_functorch/eager_transforms.py:367]: Ok my bad, I'll fix this. So all in all we need to change the type hint and to make a check in both to verify against `None`, I'll delete the test too.

Comment (zou3519) [reply torch/_functorch/eager_transforms.py:367]: yup, thank you

Comment (PierreQuinton) [reply torch/_functorch/eager_transforms.py]: I think, the new version matches our discussions.

**Review by zou3519 — APPROVED** [review]

**Comment by zou3519** [comment]
Tests are failing. But also, I don't know why we're passing a float in, that seems like a problem

**Comment by PierreQuinton** [comment]
> Tests are failing. But also, I don't know why we're passing a float in, that seems like a problem

The test failing seems to be this one:
https://github.com/pytorch/pytorch/blob/4c397e6ec6674881ea51ba67df35334065f2f5b6/test/functorch/test_eager_transforms.py#L2079

I think we could in principle change the expected error to the `ValueError` I raise. Also notes that it checks that if it is a `Tuple` then all elements needs to be `int`s, I could add this test at the begining of both `jacrev` and `jacfwd`.

What do you think @zou3519 ?

**Review by zou3519 — APPROVED** [review]
LGTM, thank you for fixing that.

**Comment by PierreQuinton** [comment]
Apparently, I missed a test with another float. Also lint failed with a line too long (even though I ran it locally). fixing those.

**Comment by PierreQuinton** [comment]
@zou3519 Sorry about this, I forgot to change the type of errors in the test. It made me realize that `ValueError` was also wrong, we check the type of `argnums` and therefore this should be `TypeError`. Fixed.

**Review by zou3519 — APPROVED** [review]

**Comment by PierreQuinton** [comment]
I'm getting bullied a bit by this unit test. It says that :
```
"jacfwd: `argnums` should be int or Tuple[int, ...], got: <class 'float'>"
"jacfwd: `argnums` should be int or Tuple[int, ...], got: <class 'float'>"
```
Now there is no visual difference so I was wondering if there was some unprintable characters, but the test only fails for `jacfwd` and not for `jacrev`, even though it is implemented exactly in the same way. So I don't know how to solve this.

File: test/functorch/test_eager_transforms.py [hunk 2]
```diff
@@ -2078,9 +2078,16 @@ def test_repeated_argnums(self, device, jacapi):
     @jacrev_and_jacfwd
     def test_float_argnums(self, device, jacapi):
         x = torch.randn(3, device=device)
-        with self.assertRaisesRegex(RuntimeError, "must be int or Tuple"):
+        name = jacapi.__name__
+        with self.assertRaisesRegex(
+            TypeError,
+            f"{name}: `argnums` should be int or Tuple[int, ...], got: <class 'float'>",
```
Comment (zou3519) [inline test/functorch/test_eager_transforms.py:2084]: Probably being bullied by the regex and how it handles special characters. Maybe try changing it to just "should be int or Tuple"?

**Comment by PierreQuinton** [comment]
@zou3519 I think I need approval for the testing.

**Comment by PierreQuinton** [comment]
@zou3519 I think this should pass the tests and be read to merge.

**Review by zou3519 — APPROVED** [review]

**Pull request closed without merging**